depends: Update OpenSSL to 1.0.2u and fix issues it introduces

patricklodder opened the pull request:
Patches OpenSSL source and the OSX build system on top of #2580, and replaces that PR.

### OpenSSL source code patch - `cbc3aae`

This patches openssl 1.0.2's usage of `secure_getenv()` to always use the fallback `OPENSSL_issetugid()` instead, to remove reliance on a higher glibc than the minimum we currently support ([2.11](https://github.com/dogecoin/dogecoin/blob/4c93783ab64ba96e60b7ea37202c072c66f3f544/contrib/devtools/symbol-check.py#L43))

The fallback code essentially does the same thing on linux as glibc's method, with the only difference that the glibc version makes decisions on startup, whereas the openssl version does the same check each time the environment is read.

See the [glibc check](https://sourceware.org/git/?p=glibc.git;a=blob;f=elf/enbl-secure.c;h=9e47526bd3e444e1a19a8ea9fd310b6f47c4db52;hb=HEAD) and [implementation](https://sourceware.org/git/?p=glibc.git;a=blob;f=stdlib/secure-getenv.c;h=a394eebcf794c1279d66e5bcb71d4b15725e6e5a;hb=HEAD), and the [openssl check]( https://github.com/openssl/openssl/blob/OpenSSL_1_0_2u/crypto/uid.c) for reference.

This patch can be removed when glibc 2.17 is the minimum version supported, i.e. for 1.21, [where the minimum glibc version is 2.17](https://github.com/dogecoin/dogecoin/blob/b262c2b3be9e1f40b362e65cea01bdd9204260dc/contrib/devtools/symbol-check.py#L42)

### OSX build system fix - `043ce9a`

Because of `make depend` magic needed inside OpenSSL not using `config.site` (only autoconf uses it, not make), `clang` was not being found when not installed inside the `$PATH`, and making the CI, which has that installed inside the `$PATH`, compile with the wrong compiler (the provided one rather than the "native" dependency from `depends/`)

I've added this selectively cherry-picked backport from Bitcoin Core (880660a) that fixates the location of `clang` and `clang++` properly, making sure that the correct pinned binary (from native depends) gets used, and not any other `clang`, i.e. something installed on the build system exposed through `$PATH`. I have removed all code surrounding the macOS native depends build from the original commit, because we do not have that facility in 1.14.

Note that there's a detailed explanation added to the `depends/Makefile` that clarifies how binary locations specified in the different `host` and `package` descriptors are being used. More changes can and imho should be made to these descriptors, but as these do not block OpenSSL 1.0.2u compilation, we can do this separately and I have not included these with this PR.

edtubbs commented:
@patricklodder, agreed `OPENSSL_issetugid` is equivalent to the glibc initializer and a reasonable replacement for `secure_getenv`.  If your Gitian checks pass, we can proceed with this PR instead.

AbcSxyZ commented:
When I run a gitian build, I'm getting the following error report :
```
ssltest.c => ../test/ssltest.c
heartbeat_test.c => ../test/heartbeat_test.c
clienthellotest.c => ../test/clienthellotest.c
sslv2conftest.c => ../test/sslv2conftest.c
dtlstest.c => ../test/dtlstest.c
bad_dtls_test.c => ../test/bad_dtls_test.c
fatalerrtest.c => ../test/fatalerrtest.c
make[2]: Leaving directory `/home/ubuntu/build/dogecoin/depends/work/build/x86_64-apple-darwin11/openssl/1.0.2-aa9f3406ed2/ssl'
making links in engines...
make[2]: Entering directory `/home/ubuntu/build/dogecoin/depends/work/build/x86_64-apple-darwin11/openssl/1.0.2-aa9f3406ed2/engines'
make[2]: Leaving directory `/home/ubuntu/build/dogecoin/depends/work/build/x86_64-apple-darwin11/openssl/1.0.2-aa9f3406ed2/engines'
making links in tools...
make[2]: Entering directory `/home/ubuntu/build/dogecoin/depends/work/build/x86_64-apple-darwin11/openssl/1.0.2-aa9f3406ed2/tools'
make[2]: Nothing to be done for `links'.
make[2]: Leaving directory `/home/ubuntu/build/dogecoin/depends/work/build/x86_64-apple-darwin11/openssl/1.0.2-aa9f3406ed2/tools'
make[1]: Leaving directory `/home/ubuntu/build/dogecoin/depends/work/build/x86_64-apple-darwin11/openssl/1.0.2-aa9f3406ed2'
make[1]: Entering directory `/home/ubuntu/build/dogecoin/depends/work/build/x86_64-apple-darwin11/openssl/1.0.2-aa9f3406ed2'
generating dummy tests (if needed)...
make[2]: Entering directory `/home/ubuntu/build/dogecoin/depends/work/build/x86_64-apple-darwin11/openssl/1.0.2-aa9f3406ed2/test'
ideatest.c => dummytest.c
md2test.c => dummytest.c
wp_test.c => dummytest.c
rc4test.c => dummytest.c
rc5test.c => dummytest.c
mdc2test.c => dummytest.c
casttest.c => dummytest.c
jpaketest.c => dummytest.c
make[2]: Leaving directory `/home/ubuntu/build/dogecoin/depends/work/build/x86_64-apple-darwin11/openssl/1.0.2-aa9f3406ed2/test'
make[1]: Leaving directory `/home/ubuntu/build/dogecoin/depends/work/build/x86_64-apple-darwin11/openssl/1.0.2-aa9f3406ed2'

Configured for darwin64-x86_64-cc.

*** Because of configuration changes, you MUST do the following before
*** building:

        make depend
make[1]: Entering directory `/home/ubuntu/build/dogecoin/depends/work/build/x86_64-apple-darwin11/openssl/1.0.2-aa9f3406ed2'
making depend in crypto...
make[2]: Entering directory `/home/ubuntu/build/dogecoin/depends/work/build/x86_64-apple-darwin11/openssl/1.0.2-aa9f3406ed2/crypto'
../util/domd: 34: ../util/domd: clang: not found
make[2]: *** [local_depend] Error 1
make[2]: Leaving directory `/home/ubuntu/build/dogecoin/depends/work/build/x86_64-apple-darwin11/openssl/1.0.2-aa9f3406ed2/crypto'
make[1]: *** [depend] Error 1
make[1]: Leaving directory `/home/ubuntu/build/dogecoin/depends/work/build/x86_64-apple-darwin11/openssl/1.0.2-aa9f3406ed2'
make: *** [/home/ubuntu/build/dogecoin/depends/work/build/x86_64-apple-darwin11/openssl/1.0.2-aa9f3406ed2/./.stamp_configured] Error 2
make: Leaving directory `/home/ubuntu/build/dogecoin/depends'
```
Look weird to just have an error with `clang not found`, I don't know if I did a wrong manipulation of if this really an error. Tried with 1.14.5-dev and were able to build (at least mac, I had a memory error for other OSes...), adding those 2 commits break with this.

AbcSxyZ commented on depends/packages/openssl.mk:
```diff
@@ -1,9 +1,10 @@
+$(package)_patches=secure_getenv.patch
 
 define $(package)_set_vars
 $(package)_config_env=AR="$($(package)_ar)" RANLIB="$($(package)_ranlib)" CC="$($(package)_cc)"
```
In the following line, variable `$(package)_cc` never defined, using clang by default.

AbcSxyZ replied:
By defining `$(package)_cc` by gcc, getting the following error : 
```
gcc: error: x86_64: No such file or directory
gcc: error: unrecognized command line option '-arch'
```

Makefile have the following flags : 
```
CC= gcc
CFLAG= -DOPENSSL_THREADS -D_REENTRANT -pipe -O2 -I/home/ubuntu/build/dogecoin/depends \
        /x86_64-apple-darwin11/include -arch x86_64 -O3 -DL_ENDIAN -Wall -DOPENSSL_IA32_SSE2 \
        -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m \
        -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DMD5_ASM -DAES_ASM -DVPAES_ASM \
        -DBSAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM
```

patricklodder replied:
Hi!

The problem is that because of `make depend`, `config.site` variables get ignored, and there's a little hack in `depends/Makefile` that sets the path using that. This will be a problem with bionic building too; the CI uses the system-installed clang for compiling openssl right now, which is worse because that will introduce further issues with the deterministic build. There's been some work on https://github.com/bitcoin/bitcoin/commit/880660acfa547558f6ef5adff6768de95e53af6e for 0.22 but that may not solve our entire problem.

Cannot solve this by switching to gcc for macOS building either (because that will not cross-compile for Mac) so I'm looking at ways to use the config.site for `make depend`. If you have a fix that does not break determinism of the build, please let us know.

AbcSxyZ replied:
I will try to implement the related issue https://github.com/bitcoin/bitcoin/pull/19683 who introduces this commit, try my luck to see if I can get something but looks tricky.

patricklodder replied:
I've already done that and am testing right now. You can see my CI run here: https://github.com/patricklodder/dogecoin/runs/3675451209

AbcSxyZ replied:
Passing where I was crashing :) On mac logs :
```
*** Because of configuration changes, you MUST do the following before
*** building:

	make depend
make[1]: Entering directory '/home/runner/work/dogecoin/dogecoin/depends/work/build/x86_64-apple-darwin11/openssl/1.0.2-467067eb724'
making depend in crypto...
make[2]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
make[2]: Entering directory '/home/runner/work/dogecoin/dogecoin/depends/work/build/x86_64-apple-darwin11/openssl/1.0.2-467067eb724/crypto'
making depend in crypto/objects...
make[3]: Entering directory '/home/runner/work/dogecoin/dogecoin/depends/work/build/x86_64-apple-darwin11/openssl/1.0.2-467067eb724/crypto/objects'
../../util/domd ../.. -MD /home/runner/work/dogecoin/dogecoin/depends/x86_64-apple-darwin11/native/bin/clang -target x86_64-apple-darwin11 -mmacosx-version-min=10.8 --sysroot /home/runner/work/dogecoin/dogecoin/depends/SDKs/MacOSX10.11.sdk -mlinker-version=253.9 -- -DOPENSSL_THREADS -D_REENTRANT -pipe -O2 -I/home/runner/work/dogecoin/dogecoin/depends/x86_64-apple-darwin11/include -arch x86_64 -O3 -DL_ENDIAN -Wall -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DMD5_ASM -DAES_ASM -DVPAES_ASM -DBSAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -I.. -I../.. -I../modes -I../asn1 -I../evp -I../../include  -DOPENSSL_NO_DEPRECATED -DOPENSSL_NO_CAMELLIA -DOPENSSL_NO_CAPIENG -DOPENSSL_NO_CAST -DOPENSSL_NO_COMP -DOPENSSL_NO_DTLS1 -DOPENSSL_NO_EC_NISTP_64_GCC_128 -DOPENSSL_NO_GMP -DOPENSSL_NO_GOST -DOPENSSL_NO_HEARTBEATS -DOPENSSL_NO_IDEA -DOPENSSL_NO_JPAKE -DOPENSSL_NO_LIBUNBOUND -DOPENSSL_NO_MD2 -DOPENSSL_NO_MDC2 -DOPENSSL_NO_RC4 -DOPENSSL_NO_RC5 -DOPENSSL_NO_RDRAND -DOPENSSL_NO_RFC3779 -DOPENSSL_NO_RSAX -DOPENSSL_NO_SCTP -DOPENSSL_NO_SEED -DOPENSSL_NO_SHA0 -DOPENSSL_NO_SSL_TRACE -DOPENSSL_NO_SSL2 -DOPENSSL_NO_SSL3 -DOPENSSL_NO_STATIC_ENGINE -DOPENSSL_NO_STORE -DOPENSSL_NO_UNIT_TEST -DOPENSSL_NO_WEAK_SSL_CIPHERS -DOPENSSL_NO_WHIRLPOOL --  o_names.c obj_dat.c obj_lib.c obj_err.c obj_xref.c
make[3]: Leaving directory '/home/runner/work/dogecoin/dogecoin/depends/work/build/x86_64-apple-darwin11/openssl/1.0.2-467067eb724/crypto/objects'
making depend in crypto/md4...
make[3]: Entering directory '/home/runner/work/dogecoin/dogecoin/depends/work/build/x86_64-apple-darwin11/openssl/1.0.2-467067eb724/crypto/md4'
```

patricklodder replied:
Yes! Though that passed before (it passed differently because it used the wrong clang) so I'm testing all gitian builds before I add the commit here (for both docker and lxc build hosts)

rnicoll commented:
I get:

```
807c2f6a394aab4dd2d3716ebd8f6354c7ee76e0b2dd459ec70ad0e1d91922a5  dogecoin-1.14.5-aarch64-linux-gnu-debug.tar.gz
575917b58a5e88641b16db08db2427126d8b7944b9c7491f80987e7dd212687f  dogecoin-1.14.5-aarch64-linux-gnu.tar.gz
8ccd53bd3bb9196c67688d84e143203255825d2afd0c79b80b98a7ca66e0bca6  dogecoin-1.14.5-arm-linux-gnueabihf-debug.tar.gz
d30dda1fc2a803dac4509973c72a4abec54849a2c965b883312bfca50c15980a  dogecoin-1.14.5-arm-linux-gnueabihf.tar.gz
7a680083c01dec27cda4e3087a42b0d63d62fd2b1d50f4a9df6dfa6927ea69ee  dogecoin-1.14.5-i686-pc-linux-gnu-debug.tar.gz
96798a3df131c50451bcc1de3696531f5fea2e8fab73338aaaa181b7aad73fe8  dogecoin-1.14.5-i686-pc-linux-gnu.tar.gz
659e9b39455a73887e8b87d059d529f702b1f7d3098cb51dee7781e05cd46c1e  dogecoin-1.14.5-x86_64-linux-gnu-debug.tar.gz
c2bd95afd59265644bdb2178c0b189d4dcbeb5c0473719c923b641522ab54bbd  dogecoin-1.14.5-x86_64-linux-gnu.tar.gz
c7cd10c11aac87eb3e838b3d314d95e25de4f9a5725d0d0811af8615d763eccc  src/dogecoin-1.14.5.tar.gz
bdd4cf63969df706a305f141c79045a5381cd33f6086d373f7e37ef44208ff1d  dogecoin-linux-1.14-res.yml
```

That look right?

rnicoll commented:
Seems to work for me, presuming the hashes above match.

rnicoll commented:
(Obviously I mean Linux version seems to work)

patricklodder commented:
For `e2083f9` my hashes, without osx, are:

```
807c2f6a394aab4dd2d3716ebd8f6354c7ee76e0b2dd459ec70ad0e1d91922a5  dogecoin-1.14.5-aarch64-linux-gnu-debug.tar.gz
575917b58a5e88641b16db08db2427126d8b7944b9c7491f80987e7dd212687f  dogecoin-1.14.5-aarch64-linux-gnu.tar.gz
8ccd53bd3bb9196c67688d84e143203255825d2afd0c79b80b98a7ca66e0bca6  dogecoin-1.14.5-arm-linux-gnueabihf-debug.tar.gz
d30dda1fc2a803dac4509973c72a4abec54849a2c965b883312bfca50c15980a  dogecoin-1.14.5-arm-linux-gnueabihf.tar.gz
7a680083c01dec27cda4e3087a42b0d63d62fd2b1d50f4a9df6dfa6927ea69ee  dogecoin-1.14.5-i686-pc-linux-gnu-debug.tar.gz
96798a3df131c50451bcc1de3696531f5fea2e8fab73338aaaa181b7aad73fe8  dogecoin-1.14.5-i686-pc-linux-gnu.tar.gz
bf61827e276475ef3210160a91ba7230f2451bc4776fb7529eaa82b102226313  dogecoin-1.14.5-win32-debug.zip
7fdf785ad7a7907259e347a0d9c2a2db89500e59deb36e52d89344e8c7b6a094  dogecoin-1.14.5-win32.zip
6448262afec61eb6f76fa82ac1c02917a894e3bd12df547ad8a9c800fe8ff517  dogecoin-1.14.5-win64-debug.zip
64d982496394d9f01c5274d171ba3a9826fef7093777af4ff02cfaca12b1be29  dogecoin-1.14.5-win64.zip
659e9b39455a73887e8b87d059d529f702b1f7d3098cb51dee7781e05cd46c1e  dogecoin-1.14.5-x86_64-linux-gnu-debug.tar.gz
c2bd95afd59265644bdb2178c0b189d4dcbeb5c0473719c923b641522ab54bbd  dogecoin-1.14.5-x86_64-linux-gnu.tar.gz
```

so that looks right. 

We'll need to do another check once the OSX patch is in though, because a different git commit will change all the checksums. I'll be more sure about its correctness tomorrow.

patricklodder commented:
Added the OSX fix, rewrote the subject line and the description of this PR and rebased on top of the latest 1.14.5-dev. This is now ready for review and gitian comparison.

patricklodder commented:
These are my LXC-host, gitian-built `sha256sum` digests for 043ce9a:

```
ca87c9e83de3a99ee4c790577b5196955d540d121b3e3a267a62cda57ea78220  dogecoin-1.14.5-aarch64-linux-gnu-debug.tar.gz
5c7c0cd0099b52542ff6e2da732e337b0ecd7327eff2d88c40343129b8fbf369  dogecoin-1.14.5-aarch64-linux-gnu.tar.gz
ca6f7ec5d7bc57794d2c80ad0a0c0b2dd0b28c4ba2f8aba763bf492852a33eac  dogecoin-1.14.5-arm-linux-gnueabihf-debug.tar.gz
5a8a1a96bf7bbcdb4063900408ea817836d1b71949116d113881432ab5c3ba49  dogecoin-1.14.5-arm-linux-gnueabihf.tar.gz
20d2fa82c7028e0da7203a2664a9a5f8b1fa93a0ed96470ecf4f2921d7f5727c  dogecoin-1.14.5-i686-pc-linux-gnu-debug.tar.gz
8cfffa04d764ba6d713d306e1526c8e359498f9f06298dbbd0f9c9cf74c611ae  dogecoin-1.14.5-i686-pc-linux-gnu.tar.gz
66f829b43d7b2cf7dc6cb24f98a6e4e2e2eb79d28ae67fc6cb12934e82163f4a  dogecoin-1.14.5-osx-unsigned.dmg
5f035042b07a49b42141728a2bdc4ae6060c6bab56671a031e6c7cec88a6ca7b  dogecoin-1.14.5-osx-unsigned.tar.gz
d22ffb590ac5a3a11a2bb8399ba3b13d540019572551163a44254afb6c14c795  dogecoin-1.14.5-osx64.tar.gz
830524e54e9f8b941859c35b7e59492eda3c0fedb14961d01df1a528efef8489  dogecoin-1.14.5-win32-debug.zip
b90132d940da9b87156d7c199caa06eeb1ff858a698c495b2345ca24e1ea5995  dogecoin-1.14.5-win32.zip
9d793f97a0785e8e466ec07ea670f1b3b4dcbb241c77c6bacbc955758741f043  dogecoin-1.14.5-win64-debug.zip
73369921e5e222b94a2839b4d98bd613d698039cc39b12626700d75aa00b7ee1  dogecoin-1.14.5-win64.zip
c58e12a933c56e925ab5ae158ef05e3f25a9d242d1af642bdce7a071c18f0a49  dogecoin-1.14.5-x86_64-linux-gnu-debug.tar.gz
fe70a0325534c85adbfe644f55182cf4da972a721d86594c43e60b2f23b8685b  dogecoin-1.14.5-x86_64-linux-gnu.tar.gz
e42014b9c0cf12451c932da2ce6b0800df59f92783b6eb75d23e13ff62cd67e3  dogecoin-1.14.5.tar.gz
```

Will confirm against a docker build too, we could use another independent set of hashes.

rnicoll commented:
```
$ sha256sum.exe dogecoin-1.14.5*
ca87c9e83de3a99ee4c790577b5196955d540d121b3e3a267a62cda57ea78220 *dogecoin-1.14.5-aarch64-linux-gnu-debug.tar.gz
5c7c0cd0099b52542ff6e2da732e337b0ecd7327eff2d88c40343129b8fbf369 *dogecoin-1.14.5-aarch64-linux-gnu.tar.gz
ca6f7ec5d7bc57794d2c80ad0a0c0b2dd0b28c4ba2f8aba763bf492852a33eac *dogecoin-1.14.5-arm-linux-gnueabihf-debug.tar.gz
5a8a1a96bf7bbcdb4063900408ea817836d1b71949116d113881432ab5c3ba49 *dogecoin-1.14.5-arm-linux-gnueabihf.tar.gz
20d2fa82c7028e0da7203a2664a9a5f8b1fa93a0ed96470ecf4f2921d7f5727c *dogecoin-1.14.5-i686-pc-linux-gnu-debug.tar.gz
8cfffa04d764ba6d713d306e1526c8e359498f9f06298dbbd0f9c9cf74c611ae *dogecoin-1.14.5-i686-pc-linux-gnu.tar.gz
66f829b43d7b2cf7dc6cb24f98a6e4e2e2eb79d28ae67fc6cb12934e82163f4a *dogecoin-1.14.5-osx-unsigned.dmg
5f035042b07a49b42141728a2bdc4ae6060c6bab56671a031e6c7cec88a6ca7b *dogecoin-1.14.5-osx-unsigned.tar.gz
d22ffb590ac5a3a11a2bb8399ba3b13d540019572551163a44254afb6c14c795 *dogecoin-1.14.5-osx64.tar.gz
a631d754fe40311e2215db183d96a7935bb438b6437b5a4bf6d41f651883e01d *dogecoin-1.14.5-win-unsigned.tar.gz
830524e54e9f8b941859c35b7e59492eda3c0fedb14961d01df1a528efef8489 *dogecoin-1.14.5-win32-debug.zip
8c5504decbd8f2fa7f5c0f4b5595f4e25091f7567ac0680689a2479e33101d84 *dogecoin-1.14.5-win32-setup-unsigned.exe
b90132d940da9b87156d7c199caa06eeb1ff858a698c495b2345ca24e1ea5995 *dogecoin-1.14.5-win32.zip
9d793f97a0785e8e466ec07ea670f1b3b4dcbb241c77c6bacbc955758741f043 *dogecoin-1.14.5-win64-debug.zip
0ea1f83944e70102331be17f9b4cf5b03c8045f0a57500ead83b21f99ece3313 *dogecoin-1.14.5-win64-setup-unsigned.exe
73369921e5e222b94a2839b4d98bd613d698039cc39b12626700d75aa00b7ee1 *dogecoin-1.14.5-win64.zip
c58e12a933c56e925ab5ae158ef05e3f25a9d242d1af642bdce7a071c18f0a49 *dogecoin-1.14.5-x86_64-linux-gnu-debug.tar.gz
fe70a0325534c85adbfe644f55182cf4da972a721d86594c43e60b2f23b8685b *dogecoin-1.14.5-x86_64-linux-gnu.tar.gz
```
I don't have `dogecoin-1.14.5.tar.gz` and I've got unsigned Windows files it looks like you don't, is that a concern?

patricklodder commented:
> I don't have `dogecoin-1.14.5.tar.gz` and I've got unsigned Windows files it looks like you don't, is that a concern?

Nope. I just didn't copy them out to be compared, but did copy out all archives, including the source tarball.

My docker hashes match my lxc hashes as well. So the gitian build is good now.
